gwcs/wcs.py

```diff
@@ -332,7 +388,502 @@ def _invert(self, *args, **kwargs):
+            return corr
+
+        # initial iteration:
+        dpix = correction(pix)
+
+        # Update initial solution:
+        pix -= dpix
+
+        # Norm (L2) squared of the correction:
+        dn = np.sum(dpix * dpix, axis=1)
+        dnprev = dn.copy()  # if adaptive else dn
+        tol2 = tolerance**2
+
+        # Prepare for iterative process
+        k = 1
+        ind = None
+        inddiv = None
+
+        # Turn off numpy runtime warnings for 'invalid' and 'over':
+        old_invalid = np.geterr()['invalid']
+        old_over = np.geterr()['over']
+        np.seterr(invalid='ignore', over='ignore')
+
+        # ############################################################
+        # #                NON-ADAPTIVE ITERATIONS:                 ##
+        # ############################################################
+        if not adaptive:
+            # Fixed-point iterations:
+            while (np.nanmax(dn) >= tol2 and k < maxiter):
+                # Find correction to the previous solution:
+                dpix = correction(pix)
+
+                # Compute norm (L2) squared of the correction:
+                dn = np.sum(dpix * dpix, axis=1)
+
+                # Check for divergence (we do this in two stages
+                # to optimize performance for the most common
+                # scenario when successive approximations converge):
+
+                if detect_divergence:
+                    divergent = (dn >= dnprev)
+                    if np.any(divergent):
+                        # Find solutions that have not yet converged:
+                        slowconv = (dn >= tol2)
+                        inddiv, = np.where(divergent & slowconv)
+
+                        if inddiv.shape[0] > 0:
+                            # Update indices of elements that
+                            # still need correction:
+                            conv = (dn < dnprev)
+                            iconv = np.where(conv)
+
+                            # Apply correction:
+                            dpixgood = dpix[iconv]
+                            pix[iconv] -= dpixgood
+                            dpix[iconv] = dpixgood
+
+                            # For the next iteration choose
+                            # non-divergent points that have not yet
+                            # converged to the requested accuracy:
+                            ind, = np.where(slowconv & conv)
+                            world = world[ind]
+                            dnprev[ind] = dn[ind]
+                            k += 1
+
+                            # Switch to adaptive iterations:
+                            adaptive = True
+                            break
+                    # Save current correction magnitudes for later:
+                    dnprev = dn
+
+                # Apply correction:
+                pix -= dpix
+                k += 1
+
+        # ############################################################
```

Comment:
If I understand correctly (and I may not), adaptive here means narrowing the iterations only to those points that haven't converged yet. If so, why not always do this?

Reply:
Because of the overhead required for this. If you know that all iterative processes (for each input world coordinate) are going to converge to 1-2 iterations within each other (that is the faster convergence is achieved in `n1` iterations and the slowest in `n2` iterations and `n2-n1` is low, then turning off `adaptive` will be faster.

Reply:
Is that based on actual timings? Sometimes it seems like it has overhead, but sometimes that is negligible.

Reply:
On actual timings on NIRCAM image that I use for testing - I do not see much difference. For HST images - see https://github.com/astropy/astropy/pull/2373#issuecomment-41541381 I do not know the origin of the differences and I do not think it is worth exploring. Maybe it just depends on the particular form of transformation, ie., SIP vs full NIRCAM transform and the weight these transformations have in the entire iterative process (SIP is "lightweight" compared to full NIRCAM and so indexing overhead is not felt that much for NIRCAM - just some hypotheses).